etl/src/error.rs

```diff
@@ -0,0 +1,898 @@
+    WithDescription(ErrorKind, &'static str),
+    /// Error with kind, static description, and dynamic detail
+    WithDescriptionAndDetail(ErrorKind, &'static str, String),
+    /// Multiple aggregated errors
+    Many(Vec<ETLError>),
+}
+
+/// Specific categories of errors that can occur during ETL operations.
+///
+/// This enum provides granular error classification to enable appropriate error handling
+/// strategies. Error kinds are organized by functional area and failure mode.
+#[derive(PartialEq, Eq, Copy, Clone, Debug)]
+#[non_exhaustive]
+pub enum ErrorKind {
+    /// Database connection failed or resource limitations
+    ///
+    /// Used for PostgreSQL connection errors (08xxx), resource errors (53xxx),
+    /// and general connectivity issues.
+    ConnectionFailed,
+    /// Query execution failed
+    ///
+    /// Used for PostgreSQL syntax errors (42xxx), BigQuery response errors,
+    /// and general SQL execution failures.
+    QueryFailed,
+    /// Source schema mismatch or validation error
+    SourceSchemaError,
+    /// Destination schema mismatch or validation error
+    ///
+    /// Used for PostgreSQL schema object not found errors (42xxx)
+    /// and destination schema mismatches.
+    DestinationSchemaError,
+    /// Missing table schema
+    MissingTableSchema,
+    /// Data type conversion error
+    ///
+    /// Used for PostgreSQL data conversion errors (22xxx), BigQuery column type mismatches,
+    /// UTF-8 conversion failures, and numeric parsing errors.
+    ConversionError,
+    /// Configuration error
+    ///
+    /// Used for BigQuery invalid metadata values and general configuration issues.
+    ConfigError,
+    /// Network or I/O error
+    ///
+    /// Used for PostgreSQL system errors (58xxx), BigQuery transport/request errors,
+    /// JSON I/O errors, and general I/O failures.
+    IoError,
+    /// Serialization error
+    ///
+    /// Used for BigQuery JSON serialization errors and data encoding failures.
+    SerializationError,
+    /// Deserialization error
+    ///
+    /// Used for JSON syntax/data/EOF errors and data decoding failures.
+    DeserializationError,
+    /// Encryption/decryption error
+    EncryptionError,
+    /// Authentication failed
+    ///
+    /// Used for PostgreSQL authentication errors (28xxx),
+    /// BigQuery authentication errors, and credential failures.
+    AuthenticationError,
+    /// Invalid state error
+    ///
+    /// Used for PostgreSQL transaction errors (40xxx, 25xxx),
+    /// BigQuery result set positioning errors, and state inconsistencies.
+    InvalidState,
+    /// Invalid data
+    ///
+    /// Used for BigQuery invalid column index/name errors,
+    /// UUID parsing failures, and malformed data.
+    InvalidData,
+    /// Data validation error
+    ///
+    /// Used for PostgreSQL constraint violations (23xxx)
+    /// and data integrity validation failures.
+    ValidationError,
+    /// Apply worker error
+    ApplyWorkerPanic,
```

Comment:
No action required in this PR, just a general observation: we should move away from catching panics and treat them as bugs to be fixed. Any panic should abort the process (panic = abort in release mode).

Reply:
Created issue internally.